common/utilities/timing.py

```diff
@@ -0,0 +1,127 @@
+import logging
+import time
+from functools import wraps
+
+
+class StopWatch:
+
+    start_time = None
+
+    def start_timer(self):
+        self.start_time = time.time()
+
+    def stop_timer(self):
+        end_time = time.time()
+        diff = end_time - self.start_time
+        return diff
+
+
+def time_method(method):
+    """
+    A method decorator that logs the time taken to execute a given method
+
+    :param method:
+    :return:
+    """
+
+    @wraps(method)
+    def invoke_method_with_timer(*args, **kwargs):
+        stopwatch = StopWatch()
+        stopwatch.start_timer()
+        try:
+            return method(*args, **kwargs)
+        finally:
+            duration = stopwatch.stop_timer()
+            duration = round(duration, 3)
+            logging.info({
+                'method': method.__name__,
+                'duration': duration
+            })
+
+    return invoke_method_with_timer
+
+
+def time_method_async(method):
+    """
+    A method decorator that logs the time taken to execute a given method. This decorator should be used to wrap
+    asynchronous methods as the standard time_method decorator does not handle async methods effectively
+    """
+
+    @wraps(method)
+    async def invoke_method_with_timer(*args, **kwargs):
+        stopwatch = StopWatch()
+        stopwatch.start_timer()
+        try:
+            return await method(*args, **kwargs)
+        finally:
+            duration = stopwatch.stop_timer()
+            duration = round(duration, 3)
+            logging.info({
+                'method': method.__name__,
```

Comment:
When logging, don't you have to have a format string as the first argument, and then params? Or is this a different way of logging?

Reply:
I wrote this originally as a format inspired by what spine does where they just pass in a dictionary which is formatted by the logging module. As Gareth mentioned though we should be trying to use human readable strings so this has now changed anyway